src/server/pkg/storage/fileset/index/index.proto

```diff
@@ -0,0 +1,35 @@
+syntax = "proto3";
+
+package index;
+option go_package = "github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index";
+
+import "github.com/pachyderm/pachyderm/src/server/pkg/storage/chunk/chunk.proto";
+
+// Op is the set of operations that can be associated with data.
+enum Op {
+  APPEND = 0;
+  OVERWRITE = 1;
+  DELETE = 2;
+}
+
+// DataOp is a data reference and an operation associated with the referenced data.
+message DataOp {
+  chunk.DataRef data_ref = 1;
+  Op op = 2;
+}
+
+// Datum is the id and size of a datum's contribution to a file. 
+// (bryce) how is this going to be used and does it make sense?
+message Datum {
+  int64 id = 1;
+  int64 size = 2;
+}
+
+// FileInfo is the data structure used in the fileset index for referencing the
+// content of a file.
+message FileInfo {
+  // The sequence of DataOps represents the state of the file.
+  repeated DataOp data_ops = 2;
+  // The sequence of datums that contributed to a file.
+  repeated Datum datums = 3;
+}
```

Comment:
Are these repeated fields supposed to always be the same length? If so I think it'd be better to put them in the same message. It makes a lot of sense to me to have a `Datum` field in `DataOp` to tell you which datum created this op.

Also this message could get quite big if there's a bunch of datums that are writing to the same file right? I think it might make more sense to store them as a stream using the functions in `pbutil` rather than as a repeated field. Because once they're stored as a repeated field we have to read the entire struct in to access any of them.

Reply:
Yeah, I was thinking it would make sense to have them in the DataOp I just have not spent too much time thinking about the reverse index yet (as you can see in my comment above). I mostly just left these here as placeholders.

The message size could definitely be a problem, and sort of helps the case for storing this additional metadata in the content of the tar stream. I'm pretty convinced it makes sense to not try and embed this in the tar head and instead just have this be the content in the index. I am going to go ahead and make the changes such that we store a stream of FileInfos in the content sections of the index.

Reply:
Actually, I think I recall why I decided not to put the Datums into DataOps. The reason is that DataOps contain both operations and references to data (chunk, offset, size). We don't want to have to store one of these for every datum over time (specifically chunk hashes and data ref hashes). When we coalesce, we will be able to compact the number of DataOps and we can view the contribution of a datum as an offset/size (and maybe just a size if they are in order) in the resulting byte stream. This combined with having a datum be represented by a 64 bit integer would mean that our reverse index would just need 16 bytes per entry.

Something I am just now realizing that would be challenging if we were concerned with having too much metadata for a file is that our APIs could become really nasty if we have to page in metadata for a single file. If the main concern with embedding the FileInfo is that the protobuf deserialization process could get wonky with a large message (but not too large to reasonably fit in memory), then it might still make sense to store the additional metadata in the content of the index.

Reply:
I'm not sure it'd be that bad to have 1 `DataOp` per datum, isn't that what we'll get anyways if each datum outputs to a single unique file?  Understood that you haven't spent a ton of time thinking about the reverse index so this is likely going to change, it might make sense to do the simple but inefficient thing for now and optimize later once we can actually test it. I think that compression is likely to save us from a lot of the ill effects of duplicated data in these messages.

Reply:
Yes, but that is the easy case and one where the scalability clearly correlates with number of datums. The case I am worried about is the high fan-out with overlap, such as map-reduce.

I am actually more concerned with the in-memory footprint rather than on-disk. Like I said above, I don't want us in a situation where we potentially have to page in metadata for a single file. Storing the datum info inline gives us about 10x memory footprint (~150 bytes per DataRef (two hashes, offset, size, op, datum id) vs 16 bytes for datum entry and a compacted set of DataRefs).